File: app/src/main/java/com/example/elsafa/MainActivity.java
package com.example.elsafa;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.Blob;
import com.google.firebase.firestore.FirebaseFirestore;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.io.ByteArrayOutputStream;

import Model.Offer.Offer;
import Model.User;

public class MainActivity extends AppCompatActivity {

    private FirebaseFirestore fStore;
    private FirebaseAuth auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        fStore = FirebaseFirestore.getInstance();
        auth = FirebaseAuth.getInstance();

        Animation animation = AnimationUtils.loadAnimation(this, R.anim.waiting_page_transition);
        TextView cafe_name = findViewById(R.id.cafe_name);
        cafe_name.startAnimation(animation);

        if (auth.getCurrentUser() == null) {
            signInAnonymously();
        } else {
            Thread waitingPage = new Thread() {
                @Override
                public void run() {
                    try {
                        sleep(3000);
                    } catch (Exception e) {
                    } finally {
                        Intent intent = new Intent(MainActivity.this, HomeActivity.class);
                        startActivity(intent);
                        finish();
                    }
                }
            };
            waitingPage.start();
        }
    }

    private void signInAnonymously() {
        auth.signInAnonymously().addOnSuccessListener(new OnSuccessListener<AuthResult>() {
            @Override
            public void onSuccess(AuthResult authResult) {
                addUserToFirebase();
            }
        });

        Intent intent = new Intent(this, PreView.class);
        startActivity(intent);
        finish();
    }

    private void addUserToFirebase() {
        User user = new User();
        user.setName("NA");
        user.setPhoneNumber(-1);
        user.setUID(auth.getUid());

        fStore.collection("Users")
                .document(auth.getUid())
                .set(user);

        fStore.collection("Users")
                .document(auth.getUid())
                .collection("Cart");

        fStore.collection("Users")
                .document(auth.getUid())
                .collection("WishList");
    }

    private void setOffer(int i) {
        Uri[] imagesURL = new Uri[]{
                Uri.parse("https://image.shutterstock.com/image-vector/special-offer-banner-vector-format-600w-586903514.jpg")
                , Uri.parse("https://addconn.com/images/limited.png")};


        String[] productsIds = new String[]{
                "xZvcdItmvKWqilQ9ikbm",
                "Lnslv4jK5o7GFrITi5vm"
        };

        Picasso.get().load(imagesURL[i]).into(new Target() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                Timestamp time = Timestamp.now();

                Offer offer = new Offer();
                offer.setProductId(productsIds[i]);
                offer.setDiscountPercentage(33);
                offer.setStartTime(time);
                offer.setEndTime(new Timestamp(time.getSeconds() + 86400, time.getNanoseconds()));
                offer.setCategoryName("Meals");

                ByteArrayOutputStream stream = new ByteArrayOutputStream();
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream);
                offer.setOfferPic(Blob.fromBytes(stream.toByteArray()));

                fStore.collection("Offers").document().set(offer);
            }

            @Override
            public void onBitmapFailed(Exception e, Drawable errorDrawable) {

            }

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {

            }
        });
    }
}

File: app/src/main/java/com/whiteside/cafe/MainActivity.kt
package com.whiteside.cafe

import android.content.Intent
import android.content.SharedPreferences
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.view.animation.AnimationUtils
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.Blob
import com.google.firebase.firestore.FirebaseFirestore
import com.squareup.picasso.Picasso
import com.squareup.picasso.Picasso.LoadedFrom
import com.squareup.picasso.Target
import com.whiteside.cafe.model.Offer
import com.whiteside.cafe.model.User
import com.whiteside.cafe.utils.Connection
import java.io.ByteArrayOutputStream

class MainActivity : AppCompatActivity() {
    private lateinit var fStore: FirebaseFirestore
    private lateinit var auth: FirebaseAuth
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initFirebase()

        val animation = AnimationUtils.loadAnimation(this, R.anim.waiting_page_transition)
        val cafeName = findViewById<TextView?>(R.id.cafe_name)
        cafeName.startAnimation(animation)
        Handler(mainLooper).postDelayed({ checkInternet() }, 3000)
    }

    private fun initFirebase() {
        if (FirebaseApp.getApps(applicationContext).isNullOrEmpty()) {
            val options = FirebaseOptions.Builder()
                .setApplicationId("1:261802668850:android:c88ecf82d1a11c7bb0dd2b")
                .setApiKey("AIzaSyD067FAA98I17gvHxINUNmjo4GeObS2DSQ")
                .setProjectId("ecommerce-55b58")
                .build()
            FirebaseApp.initializeApp(applicationContext, options, "Cafe")
        }
        fStore = FirebaseFirestore.getInstance()
        auth = FirebaseAuth.getInstance()
    }

    private fun checkInternet() {
        if (Connection.checkConnection(this)) {
            if (auth.currentUser == null) {
                signInAnonymously()
            } else {
                val intent = Intent(this@MainActivity, HomeActivity::class.java)
                startActivity(intent)
                finish()
            }
        } else {
            val intent = Intent(this@MainActivity, InternetActivity::class.java)
            startActivity(intent)
            finish()
        }
    }

    private fun signInAnonymously() {
        auth.signInAnonymously()
            .addOnSuccessListener { addUserToFirebase() }
            .addOnFailureListener { it.printStackTrace() }
        val intent = Intent(this, StartingAppPage::class.java)
        startActivity(intent)
        finish()
    }

    private fun addUserToFirebase() {
        val user = User()
        user.let {
            it.phoneNumber = "NA"
            it.uid = auth.uid!!
            it.cartItems = arrayListOf()
            it.wishList = arrayListOf()
            it.name = "unknown"
        }

        user.uid.let {
            fStore.collection("Users")
                .document(it)
                .set(user)

            fStore.collection("Users")
                .document(it)
                .collection("Cart")

            fStore.collection("Users")
                .document(it)
                .collection("WishList")
        }
    }

    fun setOffer(i: Int) {
        val imagesURL = arrayOf(
            Uri.parse("https://image.shutterstock.com/image-vector/special-offer-banner-vector-format-600w-586903514.jpg"),
            Uri.parse("https://addconn.com/images/limited.png")
        )
        val productsIds = arrayOf<String>(
            "xZvcdItmvKWqilQ9ikbm",
            "Lnslv4jK5o7GFrITi5vm"
        )
        Picasso.get().load(imagesURL[i]).into(object : Target {
            override fun onBitmapLoaded(bitmap: Bitmap?, from: LoadedFrom?) {
                val stream = ByteArrayOutputStream()


                val sh: SharedPreferences
                val time = Timestamp.now()
                val offer = Offer()
                offer.let {
                    it.productId = productsIds[i]
                    it.discountPercentage = 33f
                    it.startTime = time
                    it.endTime = Timestamp(time.seconds + 86400, time.nanoseconds)
                    it.categoryName = "Meals"
                    it.offerPic = Blob.fromBytes(stream.toByteArray())
                    it.type = 1
                    it.offerId = fStore.collection("Offers").document().id
                }
                bitmap!!.compress(Bitmap.CompressFormat.PNG, 100, stream)
                fStore.collection("Offers").document(offer.offerId).set(offer)

            }

            override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {}
            override fun onPrepareLoad(placeHolderDrawable: Drawable?) {}
        })
    }
}